blueoil/utils/smartdict.py

```diff
@@ -0,0 +1,47 @@
+# -*- coding: utf-8 -*-
+# Copyright 2018 The Blueoil Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# =============================================================================
+
+
+class SmartDict(dict):
+    def __init__(self, d=None, **kwargs):
+        super(SmartDict, self).__init__()
+        self.update(d or {}, **kwargs)
+
+    def update(self, d=None, **kwargs):
+        d = d or {}
+        d.update(kwargs)
+        for key, value in d.items():
+            self[key] = value
+
+    def __setitem__(self, name, value):
+        if isinstance(value, (list, tuple)):
+            value = [
+                self.__class__(x) if isinstance(x, dict) else x
```

Comment:
`SmartDict`? I wonder why there are both of `self.__class__` and `SmartDict`

Reply:
@a-hanamoto 
Your question is, why do you write "self.__class__(x)" instead of "SmartDict(x)"?
This is because if you inherit "SmartDict", "self.__class__" will be a subclass.

```python
class HyperSmartDict(SmartDict):
    pass

d = HyperSmartDict()
d.__class__ == HyperSmartDict
```

Reply:
Got your point. Thanks!